addons/website/models/website_visitor.py

```diff
@@ -4,139 +4,199 @@
-            visitor.page_count = len(visitor_info['page_ids'])
+            visitor.visitor_page_count = visitor_info['visitor_page_count']
+            visitor.page_count = visitor_info['page_count']
 
-    @api.depends('last_connection_datetime')
+    @api.depends('last_connections_ids')
     def _compute_time_statistics(self):
-        results = self.env['website.visitor'].search_read([('id', 'in', self.ids)], ['id', 'last_connection_datetime'])
-        mapped_data = {result['id']: result['last_connection_datetime'] for result in results}
-
+        results = self.env['website.visitor.lastconnection'].read_group([('visitor_id', 'in', self.ids)], ['visitor_id', 'connection_datetime:max'], ['visitor_id'])
+        mapped_data = {result['visitor_id'][0]: result['connection_datetime'] for result in results}
         for visitor in self:
-            last_connection_date = mapped_data[visitor.id]
-            visitor.time_since_last_action = _format_time_ago(self.env, (datetime.now() - last_connection_date))
-            visitor.is_connected = (datetime.now() - last_connection_date) < timedelta(minutes=5)
+            last_connection_datetime = mapped_data.get(visitor.id, False)
+            if last_connection_datetime:
+                visitor.last_connection_datetime = last_connection_datetime
+                visitor.time_since_last_action = _format_time_ago(self.env, (datetime.now() - last_connection_datetime))
+                visitor.is_connected = (datetime.now() - last_connection_datetime) < timedelta(minutes=5)
 
-    def _get_visitor_sign(self):
-        return {visitor.id: "%d-%s" % (visitor.id, self._get_visitor_hash(visitor.id)) for visitor in self}
-
-    @api.model
-    def _get_visitor_hash(self, visitor_id):
-        db_secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
-        return hmac.new(str(visitor_id).encode('utf-8'), db_secret.encode('utf-8'), sha256).hexdigest()
-
-    def _get_visitor_from_request(self):
+    def _get_visitor_from_request(self, with_previous_visitors=False):
         if not request:
             return None
-        visitor = self.env['website.visitor']
-        cookie_content = request.httprequest.cookies.get('visitor_id')
-        if cookie_content and '-' in cookie_content:
-            visitor_id, visitor_hash = cookie_content.split('-', 1)
-            if _consteq(visitor_hash, self._get_visitor_hash(visitor_id)):
-                return visitor.sudo().with_context(active_test=False).search([('id', '=', visitor_id)])  # search to avoid having to call exists()
-        return visitor
+        if with_previous_visitors and not request.env.user._is_public():
+            partner_id = request.env.user.partner_id
+            # Retrieve all the previous partner's visitors to have full history of his last products viewed.
+            return request.env['website.visitor'].sudo().with_context(active_test=False).search([('user_partner_id', '=', partner_id.id)])
```

Comment:
I know you will kill me ...
But can you go out this code?
It make no sense to have a method get_visitor_xxx that return more than one visitors with some params...
Anymore, it is not with_previous_visitors: -> it is only previous visitors and it is only called once with params True in eCom.

Sorry, I bet that it was a change from an other review